gnomeExtensions: include extensions for GNOME 48

honnip opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

`gnomeExtensions` set includes extensions for GNOME 46, 47, and 48.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jtojnar commented:
I guess we either need to remove the pano override, convert it to manually packaged extension, or keep GNOME 45 merged into the set.

jtojnar commented:
Also, we should probably sort the collisions by `pname` so that entries in `collisions.json` do not jump when an attrset with earlier sorted conflicting extension is unmerged.

```diff
--- a/pkgs/desktops/gnome/extensions/update-extensions.py
+++ b/pkgs/desktops/gnome/extensions/update-extensions.py
@@ -351,7 +351,7 @@ def find_collisions(
                 package_name_registry.setdefault(pname, set()).add(uuid)
     return {
         pname: sorted(uuids)
-        for pname, uuids in package_name_registry.items()
+        for pname, uuids in sorted(package_name_registry.items(), key=itemgetter(0))
         if len(uuids) > 1
     }
 
```

honnip commented:
I still don't know why the order is changing. Fixed it anyway by sorting the items..

jtojnar commented on pkgs/desktops/gnome/extensions/pano/default.nix:
```diff
@@ -0,0 +1,53 @@
+  meta = with lib; {
+    description = "Next-gen Clipboard Manager for Gnome Shell";
+    homepage = "https://github.com/oae/gnome-shell-pano";
+    license = licenses.gpl2;
```
`gpl2` is unclear, see https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206 for more details.

The correct one is `gpl2Plus`:

https://github.com/oae/gnome-shell-pano/blob/591570269c599a84a041604b35c2bb4165086c22/package.json#L6

jtojnar commented on pkgs/desktops/gnome/extensions/pano/default.nix:
```diff
@@ -0,0 +1,53 @@
+  ];
+
+  buildPhase = ''
+    glib-compile-schemas --strict schemas
```
When overriding `buildPhase`, you should also provide `runHook preBuild` and `runHook postBuild`. See <https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md> for more info.

jtojnar commented on pkgs/desktops/gnome/extensions/extensionRenames.nix:
```diff
@@ -42,6 +41,7 @@
   "true-color-window-invert@lynet101" = "true-color-window-invert";
   "volume_scroller@francislavoie.github.io" = "volume-scroller-2";
   "openweather-extension@penguin-teal.github.io" = "openweather-refined";
+  "fuzzy-clock@keepawayfromfire.co.uk" = "fuzzy-clock-2";
```
This one does not need to be preserved. The renames are here for when previously colliding package resolves the collision by changing its `pname` on e.g.o.

In this case, the extension got dropped from `gnomeExtensions` since it only supports GNOME ≤ 45 so the rename does not do anything.

jtojnar commented on pkgs/desktops/gnome/extensions/collisions.json:
Could you also please reorder the commits so that pano extraction & update is before the `gnomeExtensions` bump (so the commit does not break evaluation). Ideally, the sorting would also come before the `gnomeExtensions` so that the collisions do not jump around in that commit.

honnip committed:
gnomeExtensions.pano: package manually

It's been a while since the Pano extension has been uploaded to EGO, and it's not in gnomeExtensions set after GNOME 48. So, package it manually.

honnip committed:
gnomeExtensions/update: sort collision entries

jtojnar commented:
> I still don't know why the order is changing.

In Python, the dictionaries are ordered based on the order of insertion. Since the `package_name_registry` is built from `extensions.json`, which is sorted by extension’s numeric id on e.g.o, `fuzzy-clock` was added before `battery-time` if we include 1281 that maxes out on GNOME 45 but after if we omit 1281:

```json
, {"uuid": "Fuzzy_Clock@dallagi", "pname": "fuzzy-clock", "link": "https://extensions.gnome.org/extension/202/fuzzy-clock/", "shell_version_map": { "38": {} } }
, {"uuid": "fuzzy-clock@keepawayfromfire.co.uk", "pname": "fuzzy-clock", "link": "https://extensions.gnome.org/extension/1281/fuzzy-clock/", "shell_version_map": { "38": {}, "40": {}, "41": {}, "42": {}, "43": {}, "45": {} } }
, {"uuid": "batime@martin.zurowietz.de", "pname": "battery-time", "link": "https://extensions.gnome.org/extension/1475/battery-time/", "shell_version_map": { "40": {}, "41": {}, "42": {}, "43": {}, "44": {}, "45": {}, "46": {}, "47": {}, "48": {} } }
, {"uuid": "FuzzyClock@johngoetz", "pname": "fuzzy-clock", "link": "https://extensions.gnome.org/extension/4771/fuzzy-clock/", "shell_version_map": { "40": {}, "41": {}, "42": {}, "43": {}, "44": {}, "45": {}, "46": {}, "47": {} } }
, {"uuid": "batterytime@typeof.pw", "pname": "battery-time", "link": "https://extensions.gnome.org/extension/5425/battery-time/", "shell_version_map": { "43": {}, "44": {}, "45": {}, "46": {}, "47": {}, "48": {} } }
, {"uuid": "FuzzyClock@fire-man-x", "pname": "fuzzy-clock", "link": "https://extensions.gnome.org/extension/7056/fuzzy-clock/", "shell_version_map": { "45": {}, "46": {}, "47": {}, "48": {} } }
, {"uuid": "battery-time@eetumos.github.com", "pname": "battery-time", "link": "https://extensions.gnome.org/extension/7999/battery-time/", "shell_version_map": { "47": {}, "48": {} } }
```

honnip committed:
gnomeExtensions: include extensions for GNOME 48

honnip commented:
Ah I see, so it makes sense to sort by pname rather than the id, as we did here

jtojnar commented:
Thank you.
